### DISCUSSION TITLE
How can I make a preview model before saving to database.

### DISCUSSION THREAD

**Discussion by NaungYeHtet** [filamentphp/filament#9638] [Help]
Like a Stackoverflow question, I want to show a preview modal to user before saving data to database.

I am using `factory()->make()` method to not persist data and manually setting relations with `setRelation` method. But I have some difficulties to implement this.

1. Cannot add preview action to getFormActions function in create page to show beside `Create` action button, but it was not able to access `$get()`.
2. Cannot validate data manually before preview action. Since it is necessary to validate before preview with a view.
3. Cannot get all fields at once instead of `$get()`. It will be much easier if there is a function to get all form fields at once.

**Comment by leandrocfe** [comment]
A workaround

Add in the CreatePage
```php
protected function getCreateFormAction(): Action
{
    return parent::getCreateFormAction()
        ->submit(null)
        ->requiresConfirmation()
        ->mountUsing(fn (Form $form) => $form->fill(
            $this->data
        ))
        ->form([
            TextInput::make('your_field')
                ->disabled(),
        ])
        ->action(fn () => parent::create());
}
```

**Reply by NaungYeHtet** [reply]
Thanks for your answer @leandrocfe 

You are using confirmationModal as preview. Here is my current progress to show you more detail which I want to achieve.

I want to validate the form before preview. And also I want to show information user in a custom modal view, So I put it inside afterFormValidated. 
The modal is not appear and the action not halted when I use like this.
```
protected function getCreateFormAction(): Action
{
    return Action::make('create')
        ->submit(null)
        ->label('Preview & Create')
        ->afterFormValidated(function (Action $action)  {
            $action->halt();
            $action->modalContent(function (): View {
                $post = Post::factory()->make($this->data);

                return view('website.pages.posts.show', [
                    'data' => $this->data,
                ]);
            });
        })
        ->keyBindings(['mod+s']);
}
```

Another workaround. The problem is It lost validation if I use modalContent like this.
```
Action::make('create')
    ->submit(null)
    ->label('Preview & Create')
    ->modalContent(function (): View {
        $this->form->validate(); // but this throw validation error on the modal
        $post = Post::factory()->make($this->data);

        return view('website.pages.posts.show', [
            'post' => $post,
        ]);
    })
```

`$this->form->validate()` throws validation error but in the modal. It would do if it can show it in the form.

If I got this success, I will create another preview button.

**Reply by leandrocfe** [reply]
hum.. what about a [Wizard](https://filamentphp.com/docs/3.x/panels/resources/creating-records#using-a-wizard) with a preview step?

**Reply by NaungYeHtet** [reply]
With wizard, I found View and ViewField to implement in Preview step.
Since I want custom view, both of them are not suitable.
Can you suggest me another?

**Reply by leandrocfe** [reply]
Why did you close it? Did you solve it?

**Reply by emmanpbarrameda** [reply]
up
